docs/api-spec.md

```diff
@@ -0,0 +1,322 @@
+
+## Detailed Resource Types - v1beta1
+
+### `ArrayOrString`
+
+| Field Name  | Field Type | Requirement |
+|-------------|------------|-------------|
+| `type`      | Enum:<br>- `"string"` (default)<br>- `"array"` | REQUIRED |
+| `stringVal` | string     | REQUIRED    |
+| `arrayVal`  | []string   | REQUIRED    |
+
+### `ContainerStateRunning`
+
+| Field Name   | Field Type | Requirement |
+|--------------|------------|-------------|
+| `startedAt`* | string     | REQUIRED    |
+
+\* `startedAt` MUST be populated by the implementation, in [RFC3339](https://tools.ietf.org/html/rfc3339).
+
+### `ContainerStateWaiting`
+
+| Field Name | Field Type | Requirement |
+|------------|------------|-------------|
+| `reason`   | string     | REQUIRED    |
+| `message`  | string     | REQUIRED    |
+
+### `ContainerStateTerminated`
+
+| Field Name    | Field Type | Requirement |
+|---------------|------------|-------------|
+| `exitCode`    | int32      | REQUIRED    |
+| `reason`      | string     | REQUIRED    |
+| `message`     | string     | REQUIRED    |
+| `startedAt`*  | string     | REQUIRED    |
+| `finishedAt`* | string     | REQUIRED    |
+
+\* `startedAt` and `finishedAt` MUST be populated by the implementation, in [RFC3339](https://tools.ietf.org/html/rfc3339).
+
+### `EnvVar`
+
+| Field Name | Field Type | Requirement |
+|------------|------------|-------------|
+| `name`     | string     | REQUIRED    |
+| `value`    | string     | REQUIRED    |
+
+**NB:** All other [EnvVar](https://godoc.org/k8s.io/api/core/v1#EnvVar) types inherited from [core.v1/EnvVar](https://godoc.org/k8s.io/api/core/v1#EnvVar) and supported by the Kubernetes implementation (e.g., `valueFrom`) are **OPTIONAL** for the purposes of this spec.
+
+### `Param`
+
+| Field Name | Field Type      | Requirement |
+|------------|-----------------|-------------|
+| `name`     | string          | REQUIRED    |
+| `value`    | `ArrayOrString` | REQUIRED    |
+
+### `ParamSpec`
+
+| Field Name    | Field Type | Requirement |
+|---------------|------------|-------------|
+| `name`        | string     | REQUIRED    |
+| `description` | string     | REQUIRED    |
+| `type`        | Enum: <br>- `"string"` (default) <br>- `"array"` | REQUIRED |
+| `default`     | `ArrayOrString` | REQUIRED |
+
+### `Step`
+
+| Field Name   | Field Type | Requirement |
+|--------------|------------|-------------|
+| `name`       | string     | REQUIRED    |
+| `image`      | string     | REQUIRED    |
+| `args`       | []string   | REQUIRED    |
+| `command`    | []string   | REQUIRED    |
+| `workingDir` | []string   | REQUIRED    |
+| `env`        | `[]EnvVar` | REQUIRED    |
+| `script`     | string     | REQUIRED    |
+
+**NB:** All other fields inherited from the [core.v1/Container](https://godoc.org/k8s.io/api/core/v1#Container) type supported by the Kubernetes implementation are **OPTIONAL** for the purposes of this spec.
```

Comment:
nice, thanks for whittling this down! 

is StepState similar (i.e. based on pod status) or is this something we created ourselves?

Reply:
Yeah StepState has the same problem, but it's slightly better since there at least aren't tons of extra inherited fields. But K8s could add more any time they want, so we should spell it out defensively.